### PR TITLE
mkosi: Switch to use mkosi presets with prebuilt initrds

### PR DESCRIPTION
**Pull request by daandemeyer**
Instead of building the initrds for the mkosi images with dracut,
let's switch to using mkosi presets to build the initrd with mkosi
as well.

This commit splits up our single image build into three separate
mkosi presets:

1. The "base" preset. This image contains systemd and all its runtime
dependencies. The sole purpose of this image is to serve as a base image
for the initrd and the final image. It's also responsible for building
systemd from source with the build script. The results are installed into
the base image. Note that we install the systemd and udev packages into this
image as well to prevent package managers from overriding the systemd we built
from source with the distro packaged systemd if it's pulled in as a dependency
by another package from the initrd or final profiles.
2. The "initrd" preset. This image provides the initrd. It's trivial and does
nothing more than packaging the base image up as a zstd compressed initramfs and
adds /init and /etc/initrd-release symlinks to the image.
3. The "final" preset. This image builds on top of the base image and adds
a kernel and extra packages that are useful for testing and debugging.

We also split out the optional kernel build into a separate set of config files
that are only included if a kernel to build is actually provided.

Note that this commit doesn't really change anything about how mkosi is used.
The commands remain the same, except that mkosi will now build all the presets
in order. "mkosi summary" will show the summary of all the presets. "mkosi qemu,
boot, shell" will always boot the final preset. With "-f", all presets will be
built and the final one is booted. "-i" makes a cache of each preset.

The only thing to keep in mind is that specifying config via the mkosi CLI will
apply to each of the presets. e.g. any extra packages added with "-p" will be
installed in both the initrd and the final image. To apply local configuration
to a single preset, create a file 00-local.conf in
mkosi.presets/<profile>/mkosi.conf.d and put all the preset specific configuration
in there.

### CODE REVIEW

File: mkosi.presets/00-base/mkosi.prepare [hunk 1]
```diff
@@ -2,7 +2,10 @@
 # SPDX-License-Identifier: LGPL-2.1-or-later
 set -e
 
-if [ "$(grep '^ID=' /etc/os-release)" = "ID=\"centos\"" ] && [ "$(grep '^VERSION=' /etc/os-release)" = "VERSION=\"8\"" ]; then
+. /etc/os-release
+
+if [ "$ID" = "centos" ] && [ "$VERSION" = "8" ]; then
+    echo "abc"
```
Comment (bluca) [inline mkosi.presets/00-base/mkosi.prepare]: leftover dbg?

File: mkosi.conf.d/05-qemu-mem.conf [hunk 2]
```diff
@@ -0,0 +1,8 @@
+# SPDX-License-Identifier: LGPL-2.1-or-later
+
+[Match]
+Distribution=arch centos
```
Comment (bluca) [inline mkosi.conf.d/05-qemu-mem.conf:7]: maybe a comment on why this file only applies to centos and arch would be useful

**Comment by DaanDeMeyer** [comment]
@medhefgo So with this PR, when I try to boot CentOS 8 in qemu, it gets stuck in a boot loop, I keep getting systemd-boot@, followed by systemd-stub@, followed by the screen clearing and that keeps going on forever. It doesn't seem like we ever reach the kernel. Do you have any clue what might be causing that? I assume its got something to do with the x86 handover since only CentOS 8 actually exercises that logic if I understand correctly.

To reproduce, you can check out this branch and https://github.com/systemd/mkosi/pull/1506 and then run `mkosi -d centos -r 8 -f qemu` in the systemd repo

**Comment by DaanDeMeyer** [comment]
Note that the corresponding mkosi PR hasn't been merged yet: https://github.com/systemd/mkosi/pull/1506. We'll need to wait with merging this until that one is merged, but don't let that discourage anyone from reviewing this!

**Review by bluca — APPROVED** [review]

**Comment by DaanDeMeyer** [comment]
So I added a bunch of tracing and the handover() function in linux_efi_handover() is called but fails for an unknown reason causing the boot to restart from the beginning.

**Comment by DaanDeMeyer** [comment]
Huh I thought it'd be the zstd initramfs compression causing the issue on CentOS 8 but switching to xz compression doesn't fix the issue.

**Comment by medhefgo** [comment]
I'd investigate, but mkosi fails to build an image (both on main git branch):

```
‣   Applying presets…
‣ + systemctl --root /home/medhefgo/systemd/.mkosi.tmpv8rgoqh4/root preset-all
running_in_chroot(): Permission denied
Cannot find unit rsyslog.service.
Failed to preset unit, file "syslog.service": Link has been severed
‣ "systemctl --root /home/medhefgo/systemd/.mkosi.tmpv8rgoqh4/root preset-all" returned non-zero exit code 1.
Traceback (most recent call last):
File "/usr/lib/python3.10/runpy.py", line 196, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/lib/python3.10/runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "/home/medhefgo/mkosi/mkosi/__main__.py", line 53, in <module>
    main()
  File "/usr/lib/python3.10/contextlib.py", line 79, in inner
    return func(*args, **kwds)
  File "/home/medhefgo/mkosi/mkosi/__main__.py", line 49, in main
    run_verb(args, config)
  File "/home/medhefgo/mkosi/mkosi/__init__.py", line 2117, in run_verb
    fork_and_wait(target)
  File "/home/medhefgo/mkosi/mkosi/run.py", line 172, in fork_and_wait
    result = target()
  File "/home/medhefgo/mkosi/mkosi/__init__.py", line 2112, in target
    build_stuff(uid, gid, args, config)
  File "/home/medhefgo/mkosi/mkosi/__init__.py", line 1637, in build_stuff
    build_image(state, manifest=manifest, for_cache=False)
  File "/home/medhefgo/mkosi/mkosi/__init__.py", line 1527, in build_image
    run_preset_all(state)
  File "/home/medhefgo/mkosi/mkosi/__init__.py", line 1366, in run_preset_all
    run(["systemctl", "--root", state.root, "preset-all"])
  File "/home/medhefgo/mkosi/mkosi/run.py", line 245, in run
    raise e
  File "/home/medhefgo/mkosi/mkosi/run.py", line 232, in run
    return subprocess.run(cmdline,
  File "/usr/lib/python3.10/subprocess.py", line 526, in run
    raise CalledProcessError(retcode, process.args,
CalledProcessError: Command '['systemctl', '--root', '/home/medhefgo/systemd/.mkosi.tmpv8rgoqh4/root', 'preset-all']' returned non-zero exit status 1.
```

I suspect some low memory is triggering this, try increasing memory.

**Comment by DaanDeMeyer** [comment]
> I'd investigate, but mkosi fails to build an image (both on main git branch):

Hmm I can't reproduce, I get the same logs but the command succeeds:

```
‣ + systemctl --root /home/daandemeyer/projects/systemd/.mkosi.tmp7l8c0fhd/root preset-all
Cannot stat /proc/1/root: Permission denied
running_in_chroot(): Permission denied
Cannot find unit rsyslog.service.
Failed to preset unit, file "syslog.service": Link has been severed
```

I'm super confused because when running client side preset-all should always succeed, no clue why its failing on your side. Did you do a `ninja -C build` locally before building with mkosi?

> I suspect some low memory is triggering this, try increasing memory.

I already bumped to 4G but still got the same issue.

**Comment by DaanDeMeyer** [comment]
Tried gzip compression as well and got the same issue. Will have to compare the initrd generated by dracut and the one generated by mkosi to see if there's a difference.

**Comment by DaanDeMeyer** [comment]
The cpios look more or less the same, except that dracut has an extra initrd in the front with the microcode stuff. Maybe that's causing the issue?

**Comment by medhefgo** [comment]
Something really weird is happening.

I copied the created UKIs from main and this branch into a minimal qemu image. Trying to start either one of them with less than 4G memory will succeed in starting the kernel. The main branch will boot into initrd where it waits for the root device to show up. Your branch will panic in the kernel as it cannot find the root/initrd for some reason.
With 4G+ memory nothing happens after handover. If I enable kvm acceleration the machine resets shortly after calling handover.

`qemu-system-x86_64 -net none -drive if=pflash,format=raw,readonly=on,file=/usr/share/edk2-ovmf/x64/OVMF_CODE.fd -drive if=pflash,format=raw,file=/tmp/vars.fd -drive file=/tmp/mkosi.img,format=raw -m 8G -enable-kvm`

**Comment by DaanDeMeyer** [comment]
@medhefgo Are there any alignment requirements or such for the initramfs when doing the handover protocol? If I comment out  `Initrds=` in this final image config so it uses a dracut based initrd it boots correctly. The only difference caused by commenting out `Initrds=` is that we run kernel-install to have dracut build the initrd instead of providing it up front. So dracut must be doing something with the initrd that we are not that's causing the issue. I just have no clue what it could be.

**Comment by DaanDeMeyer** [comment]
So after more testing the initrd from this PR was broken for other reasons. I'm now at a point where both the dracut initrd and the mkosi initrd work correctly on CentOS 8 when the VM is started with < 4G of memory. When started with >= 4G of memory, they both fail to do the EFI handover, so this is an existing bug in systemd-stub or a bug in the CentOS 8 kernel.

**Comment by medhefgo** [comment]
What was the issue with the initrd?

> When started with >= 4G of memory, they both fail to do the EFI handover, so this is an existing bug in systemd-stub or a bug in the CentOS 8 kernel.

Depends on who you ask. I believe https://github.com/torvalds/linux/commit/04a7d0e15606769ef58d5cee912c5d08d93ded92 fixed it in the kernel, but we can also just relocate it ourselves for above 4G boots.

Man, I hate the handover code.

**Comment by DaanDeMeyer** [comment]
We weren't running depmod which meant no modules were being loaded in the initrd so we were not switching root and not loading virtio_console no we never got any output aside the output from the bootloader. So it looked like the same issue but was actually something completely different.

**Comment by DaanDeMeyer** [comment]
Created a separate issue for the EFI handover failure: https://github.com/systemd/systemd/issues/27472#issue-1690557889

**Comment by medhefgo** [comment]
It should work with 4G+ with #27474.

File: meson.build [hunk 3]
```diff
@@ -2494,6 +2494,10 @@ meson.add_install_script(meson_make_symlink,
                          rootlibexecdir / 'systemd',
                          rootsbindir / 'init')
 
+meson.add_install_script(meson_make_symlink,
```
Comment (bluca) [inline meson.build]: isn't this best done when building the initrd image?

**Review by bluca — APPROVED** [review]

File: meson.build (hunk 3 above)
Comment (daandemeyer) [reply meson.build]: It's relevant even when we're not building an initrd. You might want to boot into a cpio without switching root. In that case you have a cpio that's not an initrd, but you still need the /init symlink for the kernel to do the right thing. Given that we install /sbin/init already, I figure we might as well install this symlink as well.

**Comment by DaanDeMeyer** [comment]
Updated to use the official commit of the merged PR in mkosi

File: meson.build (hunk 3 above)
Comment (daandemeyer) [reply meson.build]: Hmm, dropped this commit since it might interfere with ramdisk loading.

**Pull request merged**